internal/services/dns/dns_zone_resource.go

```diff
@@ -4,377 +4,364 @@
+					return fmt.Errorf("creating/updating %s: %+v", soaRecordId, err)
+				}
+			}
 
-	id, err := zones.ParseDnsZoneID(d.Id())
-	if err != nil {
-		return err
-	}
+			metadata.SetID(id)
 
-	resp, err := zonesClient.Get(ctx, *id)
-	if err != nil {
-		if response.WasNotFound(resp.HttpResponse) {
-			d.SetId("")
 			return nil
-		}
-		return fmt.Errorf("retrieving %s: %+v", *id, err)
+		},
 	}
+}
 
-	soaRecord := recordsets.NewRecordTypeID(id.SubscriptionId, id.ResourceGroupName, id.DnsZoneName, recordsets.RecordTypeSOA, "@")
-	soaRecordResp, err := recordSetsClient.Get(ctx, soaRecord)
-	if err != nil {
-		return fmt.Errorf("retrieving %s: %+v", id, err)
-	}
+func (DnsZoneResource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			zonesClient := metadata.Client.Dns.Zones
+			recordSetsClient := metadata.Client.Dns.RecordSets
 
-	if err := d.Set("soa_record", flattenArmDNSZoneSOARecord(soaRecordResp.Model)); err != nil {
-		return fmt.Errorf("setting `soa_record`: %+v", err)
-	}
+			state := DnsZoneResourceModel{}
 
-	d.Set("name", id.DnsZoneName)
-	d.Set("resource_group_name", id.ResourceGroupName)
+			id, err := zones.ParseDnsZoneID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
 
-	if model := resp.Model; model != nil {
-		if props := model.Properties; props != nil {
-			d.Set("number_of_record_sets", props.NumberOfRecordSets)
-			d.Set("max_number_of_record_sets", props.MaxNumberOfRecordSets)
+			resp, err := zonesClient.Get(ctx, *id)
+			if err != nil {
+				if response.WasNotFound(resp.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
 
-			nameServers := make([]string, 0)
-			if s := props.NameServers; s != nil {
-				nameServers = *s
+			soaRecord := recordsets.NewRecordTypeID(id.SubscriptionId, id.ResourceGroupName, id.DnsZoneName, recordsets.RecordTypeSOA, "@")
+			soaRecordResp, err := recordSetsClient.Get(ctx, soaRecord)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", id, err)
 			}
-			if err := d.Set("name_servers", nameServers); err != nil {
-				return err
+
+			state.SoaRecord = flattenDNSZoneSOARecord(soaRecordResp.Model)
+
+			state.Name = id.DnsZoneName
+			state.ResourceGroupName = id.ResourceGroupName
+
+			if model := resp.Model; model != nil {
+				if props := model.Properties; props != nil {
+					state.NumberOfRecordSets = pointer.From(props.NumberOfRecordSets)
+					state.MaxNumberOfRecordSets = pointer.From(props.MaxNumberOfRecordSets)
+					state.NameServers = pointer.From(props.NameServers)
+				}
+				state.Tags = pointer.From(model.Tags)
 			}
-		}
 
-		if err := tags.FlattenAndSet(d, model.Tags); err != nil {
-			return err
-		}
+			return metadata.Encode(&state)
+		},
 	}
+}
 
-	return nil
+func (r DnsZoneResource) Update() sdk.ResourceFunc {
+	return r.Create()
```

Comment:
We need to write out a separate Update method because the way it is currently written will break `ignore_changes`. We need to grab the existing resource from Azure and use `metadata.ResourceData.HasChange("attribute_name")` on all top level updatable attributes. Our Guide on [creating a new resource](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-new-resource.md) has more information but also happy to answer any questions you might have when writing this out

Reply:
Thanks for the pointer, I've made a separate Update method now looking for changes on `soa_record` and `tags` as they're the only top level attributes that don't force a re-creation.
Would appreciate your feedback on this to make sure I'm on the right track